cmds/switch_root/switch_root.go

```diff
@@ -23,140 +26,171 @@ func usage() string {
 	return "switch_root [-h] [-V]\nswitch_root newroot init"
 }
 
-// littleDoctor recursively deletes everything at "path" that
-// is in the same file system as "fs".
-func littleDoctor(path string, fsType int64) error {
-	var pathFS syscall.Statfs_t
+// get device for file descriptor
+func getDev(fd int) (dev uint64, err error) {
+	var stat syscall.Stat_t
 
-	if err := syscall.Statfs(path, &pathFS); err != nil {
-		return err
+	if err := syscall.Fstat(fd, &stat); err != nil {
+		return 0, err
 	}
 
-	if int64(pathFS.Type) != fsType {
-		return nil
+	return stat.Dev, nil
+}
+
+// recursively deletes a directory and everything in it
+// works with a file descriptor, can delete files not referenceable
+// any more (e.g. after chroot)
+// does not descent into mounts
+func recursiveDelete(fd int) error {
+	parentDev, err := getDev(fd)
+	if err != nil {
+		return err
 	}
 
-	file, err := os.Open(path)
+  // the file descriptor is already open, but allocating a os.File
+	// here for it makes reading the files in the dir so much nicer
+	dir := os.NewFile(uintptr(fd), "__ignored__") // filename is completely irrelevant
```

Comment:
you can leave that comment out, the string itself already indicates that. (Go also doesn't normally have line comments like this, they usually precede the line.)